File: backend/src/wattsup/api/routes/status.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, Request

from wattsup.core.auth import require_authenticated
from wattsup.schemas.status import UpsStatus
from wattsup.services.ups import UpsManager

router = APIRouter(tags=["UPS"], dependencies=[Depends(require_authenticated)])


def manager(request: Request) -> UpsManager:
    return request.app.state.ups_manager  # type: ignore[no-any-return]


@router.get("/ups")
async def list_ups(request: Request) -> list[dict[str, str | int]]:
    units = await manager(request).list_units()
    return [
        {
            "id": unit.id,
            "name": str(unit.id),
            "description": f"{unit.server_name} · {unit.name}",
        }
        for unit in units
    ]


@router.get("/status", response_model=UpsStatus)
async def get_status(
    request: Request,
    ups: Annotated[int, Query()],
) -> UpsStatus:
    try:
        return await manager(request).status(ups)
    except LookupError as error:
        raise HTTPException(status_code=404, detail=str(error)) from error

File: frontend/src/api/status.ts
import type { UpsStatus } from "../types/status";
import { apiFetch } from "./client";

export interface UpsUnit {
  name: string;
  description: string;
}

export async function getUpsUnits(): Promise<UpsUnit[]> {
  const response = await apiFetch("/api/ups");
  if (!response.ok) throw new Error("Could not discover UPS units");
  return response.json() as Promise<UpsUnit[]>;
}

export async function getStatus(ups: string): Promise<UpsStatus> {
  const response = await apiFetch(`/api/status?ups=${encodeURIComponent(ups)}`);
  if (!response.ok) {
    throw new Error(`Status request failed (${response.status})`);
  }
  return response.json() as Promise<UpsStatus>;
}
